File: color_transfer.cpp
#include <iostream>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>


using namespace std;
using namespace cv;

void colorTransfer(const Mat& src, Mat& dst) {
	Mat labsrc, labdst;

	cvtColor(src, labsrc, COLOR_BGR2Lab);
	cvtColor(dst, labdst, COLOR_BGR2Lab);

	labsrc.convertTo(labsrc, CV_32FC3);
	labdst.convertTo(labdst, CV_32FC3);

	//计算三个通道的均值与方差
	Scalar meansrc, stdsrc, meandst, stddst;
	meanStdDev(labsrc, meansrc, stdsrc);
	meanStdDev(labdst, meandst, stddst);

	//三通道分离
	vector<Mat> /*srcsplit,*/ dstsplit;
    // split(labsrc, srcsplit);
	split(labdst, dstsplit);

	for (int i = 0; i < 3; i++) {
       dstsplit[i] -= meandst[i];
       dstsplit[i] *= (stddst[i] / stdsrc[0]);
       dstsplit[i] += meansrc[i];
    }

	//合并每个通道
	merge(dstsplit, dst);
    dst.convertTo(dst, CV_8UC1);

	//从lab空间转换到RGB空间
	cvtColor(dst, dst, COLOR_Lab2BGR);
}


int main(int argc, char** argv) {
	Mat src, dst;
	src = imread("images/ocean_day.jpg", CV_LOAD_IMAGE_COLOR);
	dst = imread("images/shoe.jpg", CV_LOAD_IMAGE_COLOR);

	colorTransfer(src, dst);
	imwrite("dst.jpg", dst);
    // imshow("src", src);
	imshow("dst", dst);
	waitKey(0);
	return 0;
}

File: color_transfer.py
#!/anaconda/bin/python

import cv2
import numpy as np


def color_transfer(src_img, dst_img):
    src_lab = np.float32(cv2.cvtColor(src_img, cv2.COLOR_BGR2LAB))
    dst_lab = np.float32(cv2.cvtColor(dst_img, cv2.COLOR_BGR2LAB))
    # src_lab = cv2.cvtColor(src_img, cv2.COLOR_BGR2LAB)
    # dst_lab = cv2.cvtColor(dst_img, cv2.COLOR_BGR2LAB)

    # Find mean and std of each channel for each image src_mean, dst_mean, src_stddev, dst_stddev
    src_mean, src_stddev = cv2.meanStdDev(src_lab)
    dst_mean, dst_stddev = cv2.meanStdDev(dst_lab)

    # Split into individual channels
    # src_channels = cv2.split(src_lab)
    dst_channels = cv2.split(dst_lab)

    print 'src_stddev[0]:', type(src_stddev), src_stddev[0]
    print 'src_mean[0]:', type(src_mean[0]), src_mean[0]

    # For each channel calculate the color distribution
    for ch in xrange(3):
        dst_channels[ch] = dst_channels[ch] - dst_mean[ch]
        dst_channels[ch] = dst_channels[ch] * (dst_stddev[ch] / src_stddev[0])
        dst_channels[ch] = np.clip(dst_channels[ch] + src_mean[ch], 0, 255)

    # TODO: mat::convertTo error here
    out_img = np.uint8(cv2.merge(dst_channels))
    return cv2.cvtColor(out_img, cv2.COLOR_LAB2BGR)


if __name__ == '__main__':
    # print "version:", cv2.__version__

    img_dir = 'images'
    src = cv2.imread('%s/%s' % (img_dir, 'ocean_day.jpg'))
    dst = cv2.imread('%s/%s' % (img_dir, 'ocean_sunset.jpg'))
    output = color_transfer(src, dst)

    cv2.imwrite('%s/%s' % (img_dir, 'output.jpg'), output)
    cv2.imshow('color transfer', output)
    cv2.waitKey(15000)
    cv2.destroyAllWindows()
